### ISSUE TITLE
Example of a CLOSURE (Function) that throws an unexpected run-time error

### ISSUE THREAD

**Issue by draegtun** [metaeducation/ren-c#605]
While reading forum post [What happens to FUNCTION! arguments and locals when the call ends?](https://forum.rebol.info/t/what-happens-to-function-arguments-and-locals-when-the-call-ends/234) I remembered there was an old Paul Graham post [Accumulator Generator](http://www.paulgraham.com/accgen.html) which included a small Rebol example (provided by @earl)...

    foo: func [n] [
        func [i] [n: n + i]
    ]

Above works correctly in Rebol 2:

    >> bar: foo 10

    >> bar 1
    == 11

    >> bar 1
    == 12

    >> bar 2
    == 14

To get this to work in R3-Alpha then we need to change first func to a `CLOS`:

    foo: clos [n] [
        func [i] [n: n + i]
    ]

However using FUNC or CLOS throws a run-time error in Ren-C...

    ;; func
    ** Script Error: n word is bound relative to context not on stack

    ;; clos
    ** Script Error: clos doesn't have RETURN: enabled for function!

Based on [Closure if obselete (trello)](https://trello.com/c/HcRtAnE4/119-epic-closure-is-obsolete-function-can-now-do-everything-closure-tried-to-offer-while-not-bearing-closures-costs) and forum post I believe the FUNC example should work (ie. *close over* N) and that we shouldn't be using CLOS/CLOSURE anymore (I assume error is an artefact from pre trello change).

**Comment by hostilefork** [comment]
When OneFunction was being first developed, the initial compromise I made was that FUNC and CLOS...as pure variants of MAKE FUNCTION!, did not put `<durable>` in the spec.

The concept was that `<durable>` would be a property you could apply to a set of variables in the frame, or all of them.  *(Similar to what I'm proposing `<has>` might do, though with that I was considering that if you wanted an argument or refinement to be durable, you would have to put it into one of the <has> "local" persistent variables...that `<has>` is just a more persistent form of `<local>`)*

Under these old and disliked-by-me rules,  you would need to say:

    foo: func [<durable> n] [
        func [i] [n: n + i]
    ]

While CLOS and FUNCTION and CLOSURE all add `<durable>` assuming you want the feature.  CLOS just apparently wasn't used by anybody so it fell victim to the not-allowed-to-return-FUNCTION! by default type checking, which is easy enough to fix with **return: [function!]**, which we might as well not fix because we want to kill it.

A motivation for why I'm bringing it up is that we've got a lot more options in the mechanics--most notably, as I said, the ability to *detect* a bound word escaping, and then create the persistent object on demand.  Previously the GC-aware handle had to be made in advance, just-in-case--every function call thus creating an object (varlist).

It appears that Rebol2 was deemed too slow to do this to every function call.  So R3-Alpha wanted to skimp on it unless you asked, so that FUNC/FUNCTION could be faster in the common case.

I think we can do a reasonable job of accomplishing what @earl and I wished for as *"closure semantics for all FUNCTION!s at acceptable cost"*.  My only question is about whether it's somehow bad to be doing this behind the user's back; as they may actually *want* arguments and locals to expire after the call, and find out about it.  I kind of like the explicit `<has>` proposal, which would use what I've in the past called "hybrid frames".

**Comment by draegtun** [comment]
> I think we can do a reasonable job of accomplishing what @earl and I wished for as "closure semantics for all FUNCTION!s at acceptable cost". My only question is about whether it's somehow bad to be doing this behind the user's back; as they may actually want arguments and locals to expire after the call, and find out about it. I kind of like the explicit <has> proposal, which would use what I've in the past called "hybrid frames".

I think the problem here is we want both!  

If i had to choose then it's "*closure semantics for all FUNCTION!s at acceptable cost"*.  If we can't achieve at an acceptable cost then I have no issue keeping CLOS/CLOSURE (ie. `func [<durable>]`) and explaining that these could leak data.

> A motivation for why I'm bringing it up is that we've got a lot more options in the mechanics--most notably, as I said, the ability to detect a bound word escaping, and then create the persistent object on demand. Previously the GC-aware handle had to be made in advance, just-in-case--every function call thus creating an object (varlist).

To clarify -  this is an (automatic) alternative to the `<has>` proposal?

**Comment by hostilefork** [comment]
> To clarify - this is an (automatic) alternative to the <has> proposal?

Just mentioning something fairly mechanical.  It used to be that when a REBVAL value cell was moved around, it was bit-copied.  e.g.

     REBVAL *some_val = ...;
     REBVAL *other_val = ...;
     *some_val = *other_val;

This meant that there was no opportunity to do any analysis on the "nature of the target cell" during a move.  But now this form of pure bit copying is disallowed *(we can enforce this, thankfully, since the code can be built with C++ and an overload of `=` can be disabled...so all you have to do is build as C++ to catch violators)*.  Every case uses `Move_Value(some_val, other_val);`, and the target cell has to be "preformatted"...you can sniff it to see what stack level it lives at, and if it's above the stack level of the frame (or indefinite lifetime, e.g. in a BLOCK!).  Action can be taken in that case--producing some longer-lived thing from the frame pointer.

So basically there's a lot of control, it's just a question of how to apply it.

**Comment by hostilefork** [comment]
For the moment, CLOSURE and `<durable>` and related ideas are being punted on.  People are advised to use `USE` if they really need the feature at the moment, as it is now an independent native:

https://forum.rebol.info/t/what-happens-to-function-arguments-and-locals-when-the-call-ends/234/9?u=hostilefork

I'm hoping that by streamlining the code and not being distracted with closure, it will be possible to build better and faster parts... that virtual binding might progress... and it will be clearer what features are needed and how to implement them.
